src/sinks/util/uri.rs

```diff
@@ -44,27 +74,107 @@ impl<'a> Visitor<'a> for UriVisitor {
     where
         E: Error,
     {
-        let uri = s.parse::<Uri>().map_err(Error::custom)?;
-        Ok(UriSerde(uri))
+        s.parse().map_err(Error::custom)
     }
 }
 
-impl From<UriSerde> for Uri {
-    fn from(t: UriSerde) -> Self {
-        t.0
+impl FromStr for UriSerde {
+    type Err = <Uri as FromStr>::Err;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        s.parse::<Uri>().map(Into::into)
     }
 }
 
 impl From<Uri> for UriSerde {
-    fn from(t: Uri) -> Self {
-        Self(t)
+    fn from(uri: Uri) -> Self {
+        match uri.authority() {
+            None => Self { uri, auth: None },
+            Some(authority) => {
+                let (authority, auth) = get_basic_auth(authority);
+
+                let mut parts = uri.into_parts();
+                parts.authority = Some(authority);
+                let uri = Uri::from_parts(parts).unwrap();
+
+                Self { uri, auth }
+            }
+        }
+    }
+}
+
+fn get_basic_auth(authority: &Authority) -> (Authority, Option<Auth>) {
+    let mut url = url::Url::parse(&format!("http://{}", authority)).unwrap();
+
+    let user = url.username();
+    if !user.is_empty() {
+        let user = percent_decode_str(user).decode_utf8_lossy().into_owned();
+
+        let password = url.password().unwrap_or("");
+        let password = percent_decode_str(password)
+            .decode_utf8_lossy()
+            .into_owned();
+
+        url.set_username("").unwrap();
```

Comment:
Could we make these `expect()`s to document why we expect them to never fail?

Reply:
These comments are good, but I was thinking you could just use `expect` to make the failure self documenting. Something like:

```
url.set_username("").expect("authority unexpectedly empty");
```

Reply:
These comments are good, but I was thinking you could just use `expect` to make the failure self documenting. Something like:

```
url.set_username("").expect("authority unexpectedly empty");
```

Reply:
Thanks! I replaced them with `expect`.